staking/staking-specification.md

```diff
@@ -0,0 +1,815 @@
+
+### 1.4 Staking Pools
+
+Staking pools can be created to leverage the weight of other stakers. For example, market makers earn liquidity rewards proportional to both their trade volume and amount of stake delegated to their pool. Delegators, in turn, can earn a portion of the pool's liquidity reward.
+
+Staking pools can also be used to increase voting power. Delegators share a portion of their vote with the pool, amplifying the pool's impact in governance over 0x.
+
+## 2 Architecture
+
+The smart contract architecture is derived from the proxy pattern, which allows state to be retained across upgrades to the logic.
+
+This system is composed of four deployed contracts:
+
+|Contract|Description |
+|--|--|
+| Staking Contract | An upgradeable/stateless contract that implements staking logic |
+|Staking Proxy|Stores staking state and delegates to the Staking Contract|
+|ZRX Vault|Securely holds staked ZRX Tokens|
+|Read-Only Proxy|Forces read-only calls to the Staking Contract|
+
+The diagram below shows how these contracts connect to each other and the broader 0x ecosystem.
+
+<p align="center"><img src="./assets/Staking%20Architecture%20-%20Basic.png" width="700" /></p>
+
+
+### 2.1 Read-Only Mode
+
+If a vulnerability is discovered in the staking contract, operations may be halted to conduct forensics:
+
+1.  The 0x Exchange contract stops charging protocol fees.
+2.  The staking contract is set to read-only mode.
+
+<p align="center"><img src="./assets/Staking%20Architecture%20-%20Read-Only.png" width="700" /></p>
+
+### 2.2 Catastrophic Failure Mode
+
+In this worst-case scenario, state has been irreparably corrupted and the staking contracts must be redeployed. Users would re-stake under the new system, at will.
+
+1. The ZRX vault is detached from the staking contract.
+2. Users withdraw their funds from the ZRX vault directly.
+
+<p align="center"><img src="./assets/Staking%20Architecture%20-%20Catastrophic%20Failure.png" width="700" /></p>
+
+## 3 Contract Migrations
+
+This section outlines steps for managing the system of smart contracts. Operations are atomically executed as a group.
+
+### 3.1 Deploying the system
+
+1. Deploy ZRX Vault.
+2. Deploy Staking Contract (address of ZRX Vault is hardcoded in this contract).
+3. Deploy Staking Proxy.
+4. Deploy Read-Only Proxy.
+5. Attach Staking Contract to Staking Proxy.
+
+### 3.2 Upgrading Staking Proxy
+
+1. Deploy new Staking Proxy.
+2. Set Staking Proxy in ZRX Vault.
+3. Set Staking Proxy in 0x Exchanges.
+
+### 3.3 Upgrading Staking Contract
+
+1. Deploy new Staking Contract.
+2. Attach Staking Contract to Staking Proxy.
+
+### 3.4 Upgrading ZRX Vault
+
+The ZRX Vault address is hardcoded in the Staking Contract and should not be upgraded. If it does need to be changed, then set the current ZRX Vault into Catastrophic Failure mode (allowing users to withdraw their ZRX) and redeploy the entire system (seciond 3.1).
```

Comment:
We could also perform this upgrade by (1) putting the current Zrx vault into catastrophic failure mode and (2) by redeploying the staking logic contract, right?

Reply:
Potentially, but if we're in catastrophic failure more then it's presumed that the state in the staking proxy is corrupted. So a full deployment would be required.